File: CODE/T4011/T4011.cpp
#define nametask "T4011"
#include <bits/stdc++.h>
#define fi first
#define se second
#define mp make_pair
#define pb push_back
#define SZ(a) (int) a.size()
#define all(a) a.begin(), a.end()
#define FOR(i, a, b) for (int i = a; i <= b; i++)
#define FOD(i, b, a) for (int i = b; i >= a; i--)

using namespace std;
typedef long long ll;
typedef pair <int, int> pi;
const int MOD = 1e9 + 7;

int k;
string L, R;
int f[55][455][500];
bool prime[500];

void add(int &x, int y){
    x += y;
    if (x >= MOD) x -= MOD;
}

int calc(int id, bool tight, int sum, int du, int sz, string &n){
    if (id >= sz){
        return ((prime[sum]) && (du == 0));
    }
    if ((!tight) && (f[id][sum][du] != -1))
        return f[id][sum][du];
    int cnt = 0;
    int maxc = (tight) ? (n[id] - '0') : 9;
    FOR(i, 0, maxc){
        int new_tight = (i == (n[id] - '0')) ? tight : 0;
        add(cnt, calc(id + 1, new_tight, sum + i, (du * 10 + i) % k, sz, n));
    }
    if (!tight) f[id][sum][du] = cnt;
    return cnt;
}

int solve(){
    memset(prime, 1, sizeof prime);
    prime[0] = prime[1] = 0;
    for (int i = 2; i * i <= 450; i++){
        if (prime[i]) for (int j = i * i; j <= 450; j += i) prime[j] = 0;
    }
    memset(f, -1, sizeof f);
    cin >> L >> R >> k;
    int sum = 0;
    int du = 0;
    FOR(i, 0, SZ(L) - 1){
        du = (du * 10 + L[i] - '0') % k;
        sum += L[i] - '0';
    }
    while (SZ(L) < SZ(R)) L = '0' + L;
    int sz = SZ(L);
    int res = (calc(0, 1, 0, 0, sz, R) - calc(0, 1, 0, 0, sz, L) + MOD) % MOD;
    if (prime[sum] && du == 0) add(res, 1);
    return res;
}

signed main()
{
    ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
    if (fopen(nametask".inp", "r")){
        freopen(nametask".inp", "r", stdin);
        freopen(nametask".out", "w", stdout);
    }
    cout << solve();
    return 0;
}

File: CODE/T4011/T4011.py
import sys

sys.stdin = open("T4011.inp", "r")
sys.stdout = open("T4011.out", "w")

def sieve(limit):
    prime = [True] * (limit + 1)
    prime[0] = prime[1] = False
    for i in range(2, int(limit**0.5) + 1):
        if prime[i]:
            for j in range(i * i, limit + 1, i):
                prime[j] = False
    return prime

def calc(id, tight, sum_digits, remainder, n, k, prime, dp):
    if id >= len(n):
        return 1 if prime[sum_digits] and remainder == 0 else 0
    
    if not tight and dp[id][sum_digits][remainder] != -1:
        return dp[id][sum_digits][remainder]
    
    max_digit = int(n[id]) if tight else 9
    count = 0
    
    for i in range(max_digit + 1):
        new_tight = tight and (i == int(n[id]))
        count += calc(id + 1, new_tight, sum_digits + i, (remainder * 10 + i) % k, n, k, prime, dp)
    
    if not tight:
        dp[id][sum_digits][remainder] = count
    
    return count

def solve():
    L, R, k = input().split()
    k = int(k)
    prime = sieve(450)
    
    sum_digits = sum(int(d) for d in L)
    remainder = int(L) % k
    
    while len(L) < len(R):
        L = '0' + L
    
    dp = [[[-1] * k for _ in range(455)] for _ in range(55)]
    result = (calc(0, True, 0, 0, R, k, prime, dp) - calc(0, True, 0, 0, L, k, prime, dp)) % (10**9 + 7)
    
    if prime[sum_digits] and remainder == 0:
        result = (result + 1) % (10**9 + 7)
    
    return result

print(solve())
